cmd/tpm-manager/tpm-manager.go

```diff
@@ -0,0 +1,507 @@
+		return false
+	}
+	for _, taint := range taints {
+		if taint.Key == TaintKey {
+			return false
+		}
+	}
+	return true
+}
+
+func loadPolicy() (err error) {
+	manager.pcrConfigs = make([]map[string]tpm.PCRConfig, 0)
+
+	if manager.pcrConfig != "" {
+		pcrdata := make(map[string]tpm.PCRConfig)
+		pcrConfig, err := ioutil.ReadFile(manager.pcrConfig)
+		if err != nil {
+			return fmt.Errorf("Unable to read valid PCR configuration %s: %v", manager.pcrConfig, err)
+		}
+		err = json.Unmarshal(pcrConfig, &pcrdata)
+		if err != nil {
+			return fmt.Errorf("Unable to parse valid PCR configuration %s: %v", manager.pcrConfig, err)
+		}
+		for pcrKey, pcrVal := range pcrdata {
+			pcrtmp := pcrVal
+			pcrtmp.Source = manager.pcrConfig
+			pcrdata[pcrKey] = pcrtmp
+		}
+		manager.pcrConfigs = append(manager.pcrConfigs, pcrdata)
+	} else if manager.pcrConfigDir != "" {
+		err = filepath.Walk(manager.pcrConfigDir, func(path string, f os.FileInfo, err error) error {
+			if f.IsDir() {
+				return nil
+			}
+			pcrConfig, err := ioutil.ReadFile(path)
+			if err != nil {
+				glog.Errorf("Unable to read PCR configuration %s: %v", path, err)
+			}
+			pcrdata := make(map[string]tpm.PCRConfig)
+			err = json.Unmarshal(pcrConfig, &pcrdata)
+			if err != nil {
+				glog.Errorf("Unable to parse valid PCR configuration %s: %v", path, err)
+			}
+			for pcr, _ := range pcrdata {
+				pcrtmp := pcrdata[pcr]
+				pcrtmp.Source = path
+				pcrdata[pcr] = pcrtmp
+			}
+			manager.pcrConfigs = append(manager.pcrConfigs, pcrdata)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		manager.pcrConfigs, err = manager.tpmhandler.GetPolicies()
+		if err != nil {
+			glog.Errorf("Unable to obtain PCR configuration: %v", err)
+			return nil
+		}
+	}
+	return nil
+}
+
+func updateAnnotations(node *api.Node, log string) {
+	if node.ObjectMeta.Annotations == nil {
+		node.ObjectMeta.Annotations = make(map[string]string)
+	}
+	node.ObjectMeta.Annotations[LogState] = log
+}
+
+func verifyNode(node *api.Node) error {
+	address, err := nodeutil.GetNodeHostIP(node)
+	if err != nil {
+		return err
+	}
+	host := fmt.Sprintf("%s:23179", address.String())
+	tpmdata, err := manager.tpmhandler.Get(host, manager.allowUnknown)
+	if err != nil {
+		invalidateNode(node)
+		return fmt.Errorf("Invalidating Node: Unable to obtain TPM data for node %s: %v", address.String(), err)
+	}
+	quote, log, err := tpm.Quote(tpmdata)
+	if err != nil {
+		invalidateNode(node)
+		return fmt.Errorf("Invalidating Node: Unable to obtain TPM quote for node %s: %v", address.String(), err)
+	}
+
+	err = tpm.ValidateLog(log, quote)
+	if err != nil {
+		invalidateNode(node)
+		return fmt.Errorf("Invalidating Node: TPM event log does not match quote for node %s", address.String())
+	}
+
+	// Don't handle this error immediately - we want to update the annotations even if validation failed
+	logstate, err := tpm.ValidatePCRs(log, quote, manager.pcrConfigs)
+	jsonlog, jsonerr := json.Marshal(logstate)
```

Comment:
may want to at least log jsonerr if != nil - hiding subtle failures makes future debugging a pain